### PR TITLE
Allow send original image if no operation is required and browser supports the format

### PR DESCRIPTION
**Pull request by surfdude75**
All images are processed even if no operation is required.
It seems a nice default since it saves bandwidth because of the use of webp format when the browser supports it. But for some applications it waste cpu and storage resources.
I added an "ifNoOperation" option in router options So you can set it to send the original image if no operation is required and the browser supports the image format. Content-type if defined based on path and if not possible it is defined by a default one set on options.
Here is how the option is:
router: { ifNoOperation: { justForward: true, imageTypeIfUnknown: 'image/jpeg' } }

### CODE REVIEW

File: lib/router/router.js [hunk 1]
```diff
@@ -109,6 +126,15 @@ p.getInfo = function(req) {
     }
   }
 
+  // forward original image if no operation on image and browser support image format
+  if (routeInfo.imageSteps.length === 0 && this.options.ifNoOperation && this.options.ifNoOperation.justForward === true) {
```
Comment (asilvas) [inline lib/router/router.js]: Please use `noop` query. This is consistent with the existing `useOriginal` query.

https://github.com/asilvas/node-image-steam/blob/master/lib/http/connect.js#L138

Comment (asilvas) [inline lib/router/router.js]: With the new option, what's the point of using isteam if you don't ever intend to process the images? Sounds like all you need is a proxy. I get wanting to do this via optional query ^ as described above.

File: lib/router/router.js [hunk 2]
```diff
@@ -173,6 +199,20 @@ p.getStepsFromObject = function(obj) {
   return steps;
 };
 
+function getImageContentType(req, routeInfo, contentTypeIfUnknown) {
+  var acceptsHeader = req.headers.accept;
+  for (var i = 0; i < imageContentTypeKeys.length; i++) {
+    var contentType = imageContentTypeKeys[i];
+    var ctInfo = imageContentTypes[contentType];
+    console.log(contentType);
```
Comment (asilvas) [inline lib/router/router.js]: plz remove

File: lib/router/router.js (hunk 1 above)
Comment (surfdude75) [inline lib/router/router.js]: I think that I missed the useOriginal query. I think it works for me. I am checking. I have large original that I need access not so often times. I use mostly resized versions, but I would like both in same endpoint.

File: lib/router/router.js (hunk 2 above)
Comment (surfdude75) [inline lib/router/router.js]: I am closing...

File: lib/router/router.js (hunk 1 above)
Comment (surfdude75) [inline lib/router/router.js]: I tested the useOriginal and it seems it just avoid the creating of the optimized original. I disable the optimized original using the following router options -> router: { originalSteps: false }
so noop query would make senses, but how could we handle the content-type?

File: lib/router/router.js (hunk 2 above)
Comment (surfdude75) [inline lib/router/router.js]: How could we handle content-type?

File: lib/router/router.js (hunk 1 above)
Comment (surfdude75) [inline lib/router/router.js]: I disabled the optimized original because I need to keep large originals. I use mostly same resized versions.

Comment (asilvas) [inline lib/router/router.js]: You could still use the new logic of this PR if you wanted, I'd just recommend adding it to query instead via `noop` flag. That way it still leverages browser `accept`.

Feel free to re-open if you still want this feature, as a query option.

Comment (surfdude75) [inline lib/router/router.js]: I am going to add the query option too. Thanks for the merge.

Comment (surfdude75) [inline lib/router/router.js]: Just added the option for noop query. I resource has no file extension it do not take effect. Do you want I sent no content-type in this case? I did this way to way to change connect.js.

File: lib/storage/fs/index.js [hunk 3]
```diff
@@ -34,6 +34,24 @@ p.fetch = function(options, originalPath, stepsHash, cb) {
   });
 };
 
+
+function checkDir(filename, cb) {
```
Comment (asilvas) [inline lib/storage/fs/index.js]: get latest from master. This is already in master.

File: lib/router/router.js (hunk 2 above)
Comment (asilvas) [inline lib/router/router.js]: I just meant to remove the `console.log`

File: lib/router/router.js [hunk 4]
```diff
@@ -109,6 +126,17 @@ p.getInfo = function(req) {
     }
   }
 
+  // forward original image if no operation on image and browser support image format
+  if (routeInfo.imageSteps.length === 0 && ((this.options.ifNoOperation && this.options.ifNoOperation.justForward === true) || routeInfo.urlInfo.query.noop === 'true')) {
```
Comment (asilvas) [inline lib/router/router.js]: I'm not seeing why we would support this as an option, since then it'd apply to all the resources. If all you need is a proxy, you should use a proxy. The query option is fine, as it's an individual opt-out.

Comment (surfdude75) [inline lib/router/router.js]: It is not applying to all resources. It just enables to apply noop=true to all requests where no operation parameter is added. It still process the operations if operation parameters are added to request url.
I am going to explain my scenario, So you can understand my need.
I have huge jpg images where the filename is a number without extension.
90% of request are for a few small resized versions. So I add operation parameter in request url to get it.
Sometimes I need to get the huge original. So I use a request url without any parameter.
But this way the image is processed and I get a huge unnecessary image in cache.
The noop could work but as my images filenames have no file extension I can not define a content-type for it. I could change to set no content-type header and trust the browser for it, but it do not sounds good.
Enabling it as an option enables to set a content-type if no file extension is found and enable to make the noop=true as the default behavior for the running instance if desired. In my case it is desirable in my running instance sine I do not want to trust endpoint user to provide the noop query to save my server resources.

Comment (asilvas) [inline lib/router/router.js]: I see. Please rename the option `noop` for consistency with query, and remove `justForward` since it's redundant with noop.

if options.noop || query.noop

**Comment by surfdude75** [comment]
Just committed the changes

File: lib/router/router.js [hunk 5]
```diff
@@ -109,6 +126,17 @@ p.getInfo = function(req) {
     }
   }
 
+  // forward original image if no operation on image and browser support image format
+  if (routeInfo.imageSteps.length === 0 && (this.options.noop || routeInfo.urlInfo.query.noop)) {
```
Comment (asilvas) [inline lib/router/router.js]: After thinking through this, it seems as `query.noop` has no value since you depend on `this.options.noop.imageTypeIfUnknown`. Please remove the query...

This is super hacky. You're basically saying that you don't want to even know the correct image format before responding to client with the untouched image. This is a big assumption that you're images are always the same format. 

I'm not a fan of this workflow but I can see your reasoning. Will merge once you remove the query param and merge latest from master.

Comment (surfdude75) [inline lib/router/router.js]: noop query works without imageTypeIfUnknown. If the image has a file extension it uses it to get content-type otherwise it just ignores the noop query if imageTypeIfUnknown is not set.
Are you sure you want I remove?
imageTypeIfUnknown seems hacky, I am not a fan too. But have have millions of files this way. Others may have too.
One next step could be read the image file header it imageTypeIfUnknown. This way noop could be fully functional anyway, even without file extension. I known JPEG is easy to check. I need to verify the others formats. I can work on it in a next step.
Your call. Let me know.

Comment (asilvas) [inline lib/router/router.js]: Pretty sure all of this could be removed if a new `format` option of `same` was added. That way the responding image would only be changed in format if the requestor does not support the image. 

Then you would just set your default options to include `format: "same"`

Comment (surfdude75) [inline lib/router/router.js]: This option could be good only if someone would like to deactive your option to use webp when possible. Not sure if some would need it. It seems work fine.
The real problem to deal here how stop unnecessary operations when we get a request without operation parameters when save system resources is more important than webp conversion gain.
the format: "same" option would affect everything.
Actually I am pretty happy with my resized images use webp when possible. I just want to disable it on orginal to save resources.

Comment (surfdude75) [inline lib/router/router.js]: maybe noop is not the name.
But it is hard to find one for it. most people is not going to understand or neither known webp optimization behind the scene

Comment (surfdude75) [inline lib/router/router.js]: maybe the option should be bandwidthSaving: { disabledToOriginals: true }

Comment (surfdude75) [inline lib/router/router.js]: it would help to market better the bandwidth optimization you have.

Comment (surfdude75) [inline lib/router/router.js]: and the hacky would be bandwidthSaving: { disabledToOriginals: { imageTypeIfunknown: 'image/jpeg' } }

Comment (asilvas) [inline lib/router/router.js]: To clarify, is all you really need is to make `useOriginal` query option, a default? If so, `useOriginal` could be added into `options` as well. It should require minimal changes.

Comment (surfdude75) [inline lib/router/router.js]: No. It seems that use original just avoid the creation of optimizedOriginal but it still process the cached version.
useOriginal can be set as a default using:   router: { originalSteps: false }
I was the first option I setup in my application other way I could not get the original file in full size.

Comment (asilvas) [inline lib/router/router.js]: If no steps are provided, the only processing would be to apply the "format" step. If your concern is that you don't want the processed image to be saved to cache, I would agree it should not be saving the processed image if you specify `useOriginal`.

So we want an option that says "if enabled, and no image steps are provided, respond back with original". I think we could achieve that by adding `useOriginal` to `options`, and making the changes to support that option. That way the format can still be processed, but it'll always serve the original if no image operations are specified. That seems reasonable.

Comment (surfdude75) [inline lib/router/router.js]: So first thing is to fix use original. Let me check here.

Comment (surfdude75) [inline lib/router/router.js]: Checking here I found the following:
- cache just avoid the use and creation of cached image but still process it with sharp.
- based on your code comment bellow. The idea of useOriginal is just related to avoid optimizedOriginal step. It would be better to avoid double mean for it.
// ignore optimized original if request explicitly opting out
What we need is to enabled disabling of webp optimization request without operation steps.
So I seems what make more sense is to rename option as bellow:
optimization: false
optimization: { originals: false }
optimization: { originals: { ct: 'image/jpeg' } }

Comment (asilvas) [inline lib/router/router.js]: No, if you specify `useOriginal`, it will not generate an optimizedOriginal image -- it will skip that step. The only processing of sharp if there are no other steps, is to change the format to the optimal format (webp for supported browsers, otherwise jpeg). Of course you can still request a specific format. If you do not want the format changed, you would new a need a new option like `autoFormat:false`.

Comment (asilvas) [inline lib/router/router.js]: Did you want to move forward with my suggestions?

It'd be to:

1. Change the behavior of `useOriginal` to not result in storing a cached image when no image steps are provided.
2. Add a new `useOriginal` option to compliment the query, which becomes the default. Would only be the default when no image steps are provided.
3. Add a new `autoFormat` option, allowing you to adhere to the format of the source image, unless the format is not supported by the requesting browser, in which case the existing fallback flow would be used.

**Comment by surfdude75** [comment]
I think it solves the problem.
I think sharp.metadata can be very fast since original is already in memory and as it parse only header files: "Fast access to image metadata without decoding any compressed image data."

File: lib/helpers/image-type.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  ctInfo = imageContentTypes[contentType];
+  if (!ctInfo) return false;
+  var regex = ctInfo.acceptRegEx;
+  if (!regex) return false;
+  var acceptsHeader = req.headers.accept;
+  return !acceptsHeader || regex.test(acceptsHeader)
+}
+
+module.exports = function (req, image, cb) {
+  var contentType = detectByFileName(image.info.path);
```
Comment (asilvas) [inline lib/helpers/image-type.js]: should never assume the file extension is correct. users often upload files with improper extensions. Please drop this check altogether.

File: lib/helpers/image-type.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  if (browserAccepts(contentType,req)) return void cb(null,contentType);
```
Comment (asilvas) [inline lib/helpers/image-type.js]: please move this down to line 63 so you're not using an unsupported type. That way it'll still auto-fix format, if the original format is not supported by the browser.

File: lib/helpers/image-type.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    pathRegEx: /\.png$/,
+    acceptRegEx: /image\/png|\*\/\*/
+  },
+  'image/jpeg': {
+    pathRegEx: /\.jpg$/,
+    acceptRegEx: /image\/jpeg|\*\/\*/
+  },
+  'image/gif': {
+    pathRegEx: /\.git$/,
+    acceptRegEx: /image\/gif|\*\/\*/
```
Comment (asilvas) [inline lib/helpers/image-type.js]: please add `webp`. It's quite common now days.

File: lib/http/connect.js [hunk 9]
```diff
@@ -5,6 +5,9 @@ var Router = require('../router');
 var Storage = require('../storage');
 var Processor = require('../processor');
 var Throttle = require('./throttle');
+var imageType = require('../helpers/image-type');
```
Comment (asilvas) [inline lib/http/connect.js]: drop this line. You're already doing `helpers.imageType`

File: lib/http/connect.js [hunk 10]
```diff
@@ -305,10 +315,13 @@ p.getHandler = function() {
...
         'Vary': 'Accept',
         'Cache-Control': 'public,max-age=31536000',
-        'Content-Length': image.buffer.length,
-        'Content-Type': reqInfo.contentType || image.contentType
+        'Content-Length': image.buffer.length
       };
 
+      var contentType = reqInfo.contentType || image.contentType
+
+      if (contentType !== 'unknown') headers['Content-Type'] = contentType;
```
Comment (asilvas) [inline lib/http/connect.js]: plz remove. we should always know content type.

File: lib/http/connect.js [hunk 11]
```diff
@@ -305,10 +315,13 @@ p.getHandler = function() {
         'Access-Control-Max-Age': '864000', // 10 days
         'Vary': 'Accept',
         'Cache-Control': 'public,max-age=31536000',
-        'Content-Length': image.buffer.length,
-        'Content-Type': reqInfo.contentType || image.contentType
+        'Content-Length': image.buffer.length
       };
 
+      var contentType = reqInfo.contentType || image.contentType
```
Comment (asilvas) [inline lib/http/connect.js]: should be able to revert this change, content type is always known for images.

File: lib/router/router.js [hunk 12]
```diff
@@ -109,6 +110,12 @@ p.getInfo = function(req) {
     }
   }
 
+  // forward original image if no operation on image and useOriginal query or option
+  if (routeInfo.imageSteps.length === 0 && (this.options.useOriginal || routeInfo.urlInfo.query.useOriginal)) {
```
Comment (asilvas) [inline lib/router/router.js]: explicit value check: `routeInfo.urlInfo.query.useOriginal === 'true'`

File: lib/http/connect.js [hunk 13]
```diff
@@ -119,6 +122,10 @@ p.getHandler = function() {
         });
       }, $this.options.stepTimeout)],
       optimizedOriginalImage: ['requestedImage', async.timeout(function optimizedOriginalImage(results, cb) {
+        if (reqInfo.useOriginal) {
```
Comment (asilvas) [inline lib/http/connect.js]: should be: `reqInfo.useOriginal || reqInfo.urlInfo.query.useOriginal === 'true'`

to account for cases of no image steps use of original (where you're forwarding content type), and when query param useOriginal is specified and there are image steps. The idea is that you can still perform image operations on the original, when desired.

File: lib/http/connect.js [hunk 14]
```diff
@@ -199,10 +213,6 @@ p.getHandler = function() {
           // rare chance they're the same, but if so no sense in requesting the same thing
           return void cb();
         }
-        if (reqInfo.urlInfo.query.useOriginal === 'true') {
```
Comment (asilvas) [inline lib/http/connect.js]: please restore this line.

Comment (surfdude75) [inline lib/http/connect.js]: This condition is in the begin now.

File: lib/http/connect.js [hunk 15]
```diff
@@ -199,10 +212,6 @@ p.getHandler = function() {
...
```
Comment (asilvas) [inline lib/http/connect.js]: Please restore this. For use when original is requested for processed requests.

Comment (asilvas) [inline lib/http/connect.js]: Looks like this is the last fix, then should be good.

Comment (surfdude75) [inline lib/http/connect.js]: Hi,

Just committed the change.

Aloha,

Rafael Sobral

**Comment by asilvas** [comment]
merged and published as `0.32.0`. Thanks!

**Comment by surfdude75** [comment]
Thanks for the patience and time.

Aloha,

Rafael Sobral

**Comment by cnhuye** [comment]
Hello,  is this option works right now ?
I'm using 0.51.1, in electron, just for local images.
I want render local original image without any process, so it can be faster.
But i found without any router option, the local image still be compressed to 2560x1707.
So i tried this option:
    "router": {
      ifNoOperation: { justForward: true, imageTypeIfUnknown: 'image/jpeg' }
    },
It seems don't work.

How can I use this option, or how can i response a local image without process?

**Comment by cnhuye** [comment]
I found "useOriginal=true" option is nearly what I want. it's really fast.
the only problem is "useOriginal" makes the picture don't use auto-rotate based on orientation.

What's the best solution to do this "fast & auto-rotate"   ?

**Comment by asilvas** [comment]
@cnhuye if you feel this is a bug please open a new issue. But being you're not using isteam as customer facing, and have no need for caching (only using useOriginal), you're probably better off just using `sharp` directly so you can script exactly how you like.
